src/cirrus/lib/process_payload.py

```diff
@@ -198,55 +219,90 @@ def __call__(self, wfem) -> str | None:
+                # exception response.
+                db_state = (
+                    StateEnum(
+                        e.response["Item"]["state_updated"]["S"].split("_")[0],
+                    )
+                    if "Item" in e.response
+                    else StateEnum.PROCESSING
+                )
+
+                wfem.skipping(
+                    self["id"],
+                    state=db_state,
+                    payload_url=url,
+                    message=f"state before claim attempt was {previous_state}",
+                )
                 return None
+            self._fail_and_raise(wfem, e, url)
+
+        # invoke step function
+        self.logger.debug("Running Step Function %s", execution_arn)
+        started_sfn = False
+        try:
+            get_client("stepfunctions").start_execution(
+                stateMachineArn=state_machine_arn,
+                name=execution_name,
+                input=json.dumps(self.get_payload()),
+            )
+            started_sfn = True
+        except ClientError as e:
             if e.response["Error"]["Code"] == "StateMachineDoesNotExist":
                 # This failure is tracked in the DB and we raise an error
                 # so we can handle it specifically, to keep the payload
                 # falling through to the DLQ and alerting.
                 logger.error(e)
                 wfem.failed(self["id"], str(e), payload_url=url)
                 raise TerminalError() from e
-            raise
-        except Exception as e:
-            # This case should be like the above, except we don't know
-            # why it happened. We'll be conservative and not raise a
-            # terminal failure, so it will get retried in case it was
-            # a transient failure. If we find terminal failures handled
-            # here, we should add terminal exception handlers for them.
-            msg = f"failed starting workflow ({e})"
-            self.logger.exception(msg)
-            wfem.failed(self.get("id", "missing"), msg, payload_url=url)
-            raise
+            if e.response["Error"]["Code"] != "ExecutionAlreadyExists":
+                self._fail_and_raise(wfem, e, url)
+            # Let ExecutionAlreadyExists pass, to try setting PROCESSING
+
+        try:
+            # add execution to DynamoDB record
+            wfem.started_processing(
+                self["id"],
+                execution_arn=execution_arn,
+                payload_url=url,
+            )
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                db_state = e.response.get("Item", "dbitem not found in response")
+                wfem.skipping(
+                    self["id"],
+                    state=StateEnum.PROCESSING,
+                    payload_url=url,
+                    message=(
+                        "started stepfunction, but could not set processing "
```

Comment:
I might need to talk this over and understand how we get here exactly. But my concern is this: in the exception handler above we potentially don't start a step function, though we still try to set it to processing to communicate that it has started.

In the case that we find it is already processing, then the conditional check fails, right? That puts us into this block? Or I guess looking at the statedb code it means the state was not claimed.

In any case, are we actually ever hitting this block if we are actually skipping something? Not starting a step function because it already exists and then finding we can't update the state either is effectively a race between two duplicated payloads being processed in parallel process executions? I guess maybe when I think about it that way, then yeah, we do have a payload we are skipping, but it seems like it's not "started stepfunction, but could not set processing". Is there another case that could result in this code path?

Maybe the problem is we aren't tracking the state as to whether or not we successfully started a step function. Perhaps if we tracked that (via some flag var) we'd have an easier job of ensuring the proper message and message state?

Maybe my confusion could also be expressed with the question: is the case where a step function starts actually a "skipping" event? Maybe a controversial question, but do we even emit an event in this case: we knew we should have started a step function, that is it _was_ in the claimed state, but now it is not; can we assume whatever else won the race and changed the state means that the event we expected was already emitted?

Reply:
> I might need to talk this over and understand how we get here exactly. But my concern is this: in the exception handler above we potentially don't start a step function, though we still try to set it to processing to communicate that it has started.

The process lambda is trying to handle two conditions:

1. the indeterminate state restart after a lambda timeout.
2. an actual race between two lambdas.

The latter condition may not actually be possible, though I'm not certain.
 
> In the case that we find it is already processing, then the conditional check fails, right? That puts us into this block? Or I guess looking at the statedb code it means the state was not claimed.

As implemented currently, this conditional failing means the payload is either not claimed or not the execution arn we expect.  The latter feels like it could only happen in the direct race condition.
 
> ... Not starting a step function because it already exists and then finding we can't update the state either is effectively a race between two duplicated payloads being processed in parallel process executions? 

Yes, I believe that to be correct.

> ... I guess maybe when I think about it that way, then yeah, we do have a payload we are skipping, but it seems like it's not "started stepfunction, but could not set processing". Is there another case that could result in this code path?
> Maybe the problem is we aren't tracking the state as to whether or not we successfully started a step function. Perhaps if we tracked that (via some flag var) we'd have an easier job of ensuring the proper message and message state?

The variable `started_sfn` is such a variable, and lets us have different messages in each case.  Only the direct race leads to one lambda starting the execution, and the other getting to update the database.  Ignoring the `different execution arn` condition for the moment, we could update the `started_sfn==True` message to say `"direct race condition detected: started stepfunction, but database already updated "`. And leave the `started_sfn==False` message as it is.

 

> Maybe my confusion could also be expressed with the question: is the case where a step function starts actually a "skipping" event? Maybe a controversial question, but do we even emit an event in this case: we knew we should have started a step function, that is it _was_ in the claimed state, but now it is not; can we assume whatever else won the race and changed the state means that the event we expected was already emitted?

Maybe this one is one we should try to walk through in person. The `wfem.skipping` function generates `ALREADY_*` event messages.  The main purpose of these is to avoid listeners from submitting a payload, and never hearing back from cirrus (deadlock).  Note that now with the no-cost `ReturnValuesOnConditionCheckFailure` parameter set, we get back the contents of the state DB, so we could inspect that to further pin down what went wrong.

Reply:
Oh gosh, I totally missed `started_sfn`. 🤦 I guess that's what I get for trying to review at the end of a long day.

In any case, we've discussed, this is good. We just want to rename the `skipping` function to be something more obvious (`is_already` or whatever). That rename is tracked on a different comment so I'll resolve this one.